design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+
+## Value/Return
+
+See [KLIP 15](./klip-15-new-api-and-client.md).
+
+## Public APIS
+
+The following subsections describe the methods of the `Client` interface:
+```
+public interface Client {
+    ...
+}
+```
+
+### Constructors
+```
+  static Client create(ClientOptions clientOptions) {
+    return new ClientImpl(clientOptions);
+  }
+
+  static Client create(ClientOptions clientOptions, Vertx vertx) {
+    return new ClientImpl(clientOptions, vertx);
+  }
+```
+
+The Java client will be implemented as a Vert.x HttpClient. We expose a constructor that allows users to provide their own `Vertx` instance
+in order to take advantage of a shared connection pool and other properties if desired.
+
+`ClientOptions` will initially be as follows:
+```
+public interface ClientOptions {
+
+  ClientOptions setHost(String host);
+
+  ClientOptions setPort(int port);
+
+  ClientOptions setUseTls(boolean useTls);
+
+  ClientOptions setUseClientAuth(boolean useClientAuth);
+
+  ClientOptions setTrustStore(String trustStorePath);
+
+  ClientOptions setTrustStorePassword(String trustStorePassword);
+
+  ClientOptions setKeyStore(String keyStorePath);
+
+  ClientOptions setKeyStorePassword(String keyStorePassword);
+
+  ClientOptions setBasicAuthCredentials(String username, String password);
+
+  String getHost();
+
+  int getPort();
+
+  boolean isUseTls();
+
+  boolean isUseClientAuth();
+
+  boolean isUseBasicAuth();
+
+  String getTrustStore();
+
+  String getTrustStorePassword();
+
+  String getKeyStore();
+
+  String getKeyStorePassword();
+
+  String getBasicAuthUsername();
+
+  String getBasicAuthPassword();
+
+  ClientOptions copy();
+
+  static ClientOptions create() {
+    return new ClientOptionsImpl();
+  }
+}
+```
+
+We can always add additional configuration options later. We may also wish to expose the Vert.x `HttpClientOptions` for
+advanced users that wish to provide custom configs.
+
+### Transient queries -- Streaming
+
+The `Client` interface will provide the following methods for streaming the results of a transient (push or pull) query:
+```
+  /**
+   * Execute a query (push or pull) and receive the results one row at a time.
+   *
+   * @param sql statement of query to execute.
+   * @return query result.
+   */
+  CompletableFuture<QueryResult> streamQuery(String sql);
+
+  /**
+   * Execute a query (push or pull) and receive the results one row at a time.
+   *
+   * @param sql statement of query to execute.
+   * @param properties query properties.
+   * @return query result.
+   */
+  CompletableFuture<QueryResult> streamQuery(String sql, Map<String, Object> properties);
+```
+where `QueryResult` is as follows:
+```
+/**
+ * The result of a query (push or pull), streamed one row at time. Records may be consumed by either
+ * subscribing to the publisher or polling (blocking) for one record at a time. These two methods of
+ * consumption are mutually exclusive; only one method may be used (per QueryResult).
```

Comment:
The suggestion here seems more confusing and complex than the original API.

There is plenty of precedent for mixing sync and async in the same API - most messaging clients work this way. Also CompletableFuture itself works this way (it has a get method).

I think if we split this out into different APIs it's going to leave the user scratching their head and lead to a poorer user experience.

Reply:
I suppose that makes sense, I don't feel that strongly about it given the noted precedent

Reply:
To me, the usage of the same return object but only allowing to either subscribe or poll from it is weird. It might be subjective, but in the end `QueryResult` becomes stateful: when returned, the state is "empty" (not determined), and when a user calls either `subscribe` or `poll` the state is set and cannot be changed any longer. Sound like bad UX to me. I am sure that most users know if they want to poll or subscribe when they submit the query and thus they should just use a corresponding method and get a corresponding result object back.

The current `streamQuery()` method can be used for push and pull queries anyway (making the prefix "stream" overloaded/ambiguous btw) and returns a `QueryResult` while `executeQuery()` returns `List<Row>` (also for push and pull queries). So both method do the same thing: they execute a pull or push query.

The difference is actually "weird" from a user point of view because the former "streams" (ie, incrementally transfers the result) to the client while the later first fetches the whole result and afterward returns (what does only work for finite query results anyway). However, this difference seems to be an implementation detail -- the client should make this decision internally. (Also, for infinite query result `executeQuery()` cannot be used.)

The user really only cares if she polls the result or if she subscribes to the result. Thus, the simplest thing seems to be keep `executeQuery()` and use it to let people `poll()` and keep `streamQuery` (or maybe rename to `subscribeQuery()` to avoid the naming ambiguity?) and let people subscribe. If either of both fetches the full result and afterward returns, or if it gets the result incrementally would become an implementation detail. Also, both method can be used for all query types what make it easier to users. For infinite result, they always need to "stream" the result, and for finite result, the client can pick if it's "streams" or "fetches upfront". This decision could maybe exposed on the result object to give people an idea about potential "blocking" behavior for "fetching upfront"

Just my 2ct.

Reply:
>> Thus, the simplest thing seems to be keep executeQuery() and use it to let people poll() 

I don't follow what you mean. ExecuteQuery() returns the results asynchronously as a List, there is no need to poll() anything here.

Reply:
>> keep streamQuery (or maybe rename to subscribeQuery() to avoid the naming ambiguity?) and let people subscribe

I also don't follow this. Some folks prefer a synchronous access pattern (poll). How would they do this if we only provide an async API?

Reply:
I think @mjsax is proposing we switch from
```
streamQuery(); -- can either (1) subscribe to result or (2) poll
executeQuery(); -- wait and receive results in a single batch
```
(as proposed in the KLIP) to
```
subscribeQuery(); -- subscribe to result stream
executeQuery(); -- can either (1) poll (receive result messages one at a time) or (2) wait and receive all result messages in a single batch 
```
In other words, the existing proposal splits the methods for receiving query results based on streaming vs. batch, whereas Matthias's proposal splits the methods based on async vs. sync.

Reply:
What @vcrfxia says.

Reply:
Maybe I've misunderstood the suggestion, but it makes no sense to me:
1. If subscribeQuery() doesn't allow you to poll the results, how will a user who doesn't like using the async API stream the results of a query from one place to another? They will be forced to get all the results in memory in a single List with executeQuery first.
2. executeQuery() waits for all results to arrive and returns them in a List. There is no need to poll or subscribe anything, as you already have the results.

Reply:
When you say "how will a user who doesn't like using the async API stream the results" what do you exactly mean by "stream the result"? Are you talking about the internal transport mechanism how data is send from the server to the client? For this case, my answer is, the user does not make this decision but the client/server make the decision about how data is transferred. The use should not be concerned with an implementation detail like that.

> They will be forced to get all the results in memory in a single List with executeQuery first.

The proposal is to change the return type of `executeQuery` to return a `QueryResult2` (just a random placeholder to distinguish it from `QueryResult` `streamQuery()` or `subscribeQuery()` returns) on which the user can either call `poll()` or `getAll()` and `getAll` would return the `List` (if the result if finite and otherwise throw an exception).

Reply:
I think it would help if you illustrated your suggestion with some method signatures as I don't understand the suggestion yet.

Reply:
I did not sketch out the details... Something like this (just made up some "random" names):
```
CompletableFuture<ResultMetadata> streamQuery(String query, Callback callback);

ResultMetadata // return error code and/or other metadata; not used to access the result Rows

// triggered for each Row in the result
interface Callback {
  // triggered async when a new row is available
  void handle(Row row);
}

CompletableFuture<QueryResult> executeQuery(String query);

interface QueryResult extends ResultMetadata {
  // fetch one Row from the result
  Row poll(Duration timeout); // blocking

  // fetch all remaining rows (ie, previously returned rows from `poll()` would not be contained
  // -> if result is infinite, throw exception
  // after `getAllRemaining()` was called, `poll()` would always return `null`
  List<Row> getAllRemaining() throws IllegalStateException // blocking; no timeout?
}
```

Something along those lines. The idea is really to separate async subscriptions (via `streamQuery`) from  sync data polling (via `executeQuery`). The API does no imply any restriction if a push/pull query is issued, ie, both can be issue via each methods. Especially for push queries either async callbacks or sync polling are expected patterns (`getAllRemaining` would only work for some corner cases). For pull queries, the async callback are less likely to be used (still allowed), while sync polling (either one by one via `poll()` or getting all rows via `getAllRemaining()`) should be the dominant usage pattern.

Reply:
>> List<Row> getAllRemaining()

But, getAllRemaining() can't be blocking. (In the original API it returned CompletableFuture<List<Row>>)

People who work in an async way also want to execute pull queries (the typical use case) asynchronously.

Whether you want to use stream query results or retrieve them all in one go and whether you prefer a synchronous or asynchronous API are orthogonal concerns.

Also, note that any CompletableFuture result can be returned synchronously (if you prefer a sync approach) by calling get() on it :)

Reply:
> But, getAllRemaining() can't be blocking. (In the original API it returned CompletableFuture<List>)

Maybe we can just return plain `QueryResult` for `executeQuery` instead and return `CompletableFuture<List>` from `getAllRemaining()`?

> Whether you want to stream query results/retrieve them all in one go and whether you prefer a synchronous/asynchronous API are orthogonal concerns.

I agree. And there are are multiple ways to design the API around it (it's basically 4 combinations but I don't think we would want to have 4 methods?) and I think that "sync/async" is the less important one while how one "retrieves" the result is more important and the API should put subscription vs pull int the focus.

Or maybe we should have 3 methods (because in the end there are 3 ways how one retrieves the result: via a callback, via polling, via getAll). Using a callback is natively async, and polling is something in between (a single call to `poll()` is blocking but might timeout to unblock again). Last, `getAll()` can be async or sync, but is quite different to a callback or polling that are "one by one".

So maybe we could have:
```
// as currently proposed; only usable for finite results
CompletableFuture<List<Row>> executeQuery(String query);

// split the current `streamQuery()` into two methods:

CompletableFuture<QueryMetadata> subscribeQuery(String query, Consumer callback);
CompletableFuture<QueryResult> query(String query); // for polling via `QueryResult`
```

(The names are not great... -- just for illustrative purpose.)